controllers/topology/desired_state.go

```diff
@@ -298,7 +382,96 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, machineDeployme
 	desiredMachineDeploymentObj.Spec.Replicas = machineDeploymentTopology.Replicas
 
 	desiredMachineDeployment.Object = desiredMachineDeploymentObj
-	return desiredMachineDeployment, nil
+	return desiredMachineDeployment, upgraded, nil
+}
+
+// computeMachineDeploymentVersion calculates the version of the desired machine deployment.
+// The version if calculated using the the state of the current machine deployments, the current control plane
+// and the version defined in the topology.
+func computeMachineDeploymentVersion(s *scope.Scope, desiredControlPlaneState *scope.ControlPlaneState, currentMDState *scope.MachineDeploymentState, allowUpgrade bool) (string, bool, error) {
+	desiredVersion := s.Blueprint.Topology.Version
+	// If creating a new machine deployment, we can pick up the desired version
+	// Note: We are not blocking the creation of new machine deployments when
+	// the control plane or any of the machine deployments are upgrading/scaling.
+	// TODO: Double check if it is okay to add a new machine deployment if the control plane is upgrading.
+	if currentMDState == nil || currentMDState.Object == nil {
+		return desiredVersion, false, nil
+	}
+
+	// get the current version of the machine deployment
+	currentVersion := *currentMDState.Object.Spec.Template.Spec.Version
+
+	// return early if we are not allowed to upgrade the machine deployment
+	if !allowUpgrade {
+		return currentVersion, false, nil
+	}
+
+	// return early if the currentVersion is already equal to the desiredVersion
+	// no further checks required.
+	if currentVersion == desiredVersion {
+		return currentVersion, false, nil
+	}
+
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		// We dont have a control plane at this time but we have a machine deployment. (Is this possible? Manually deleted?).
+		// Do not perform any machine deployment upgrade in this case. Return the current version of the machine deployment.
+		return currentVersion, false, nil
+	}
+
+	// If the current control plane is upgrading, then do not pick up the desiredVersion yet.
+	// Return the current version of the machine deployment. We will pickup the new version after the control
+	// plane is stable.
+	cpUpgrading, err := contract.ControlPlane().IsUpgrading(s.Current.ControlPlane.Object)
```

Comment:
Not sure if it's a good idea but we could add a `IsStable` func which calls IsUpgrading and IsScaling internally and returns the correct errors. This would simplify the code in computeControlPlaneVersion and in computeMachineDeploymentVersion (if I didn't miss anything)

(also not sure if we already have a clean way in the contract to check if the control plane support replicas)

Reply:
I am more inclined to keep the 2 functions separate so that the code explicitly covers the edge cases and has comments around each case. I agree that it makes it more verbose.

I dont think there is a good way to detect from the control plane contract if it supports replicas. From these [comments here](https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/topology/internal/contract/controlplane.go#L59-L60) it looks like just checking if the field is present on the unstructured object is not enough to determine if it supports replicas.

One work around for this would be to have a function like `IsStable(checkReplicas bool)` (use an argument to determine if we need to also check for scaling).

On a related note, I am also wondering if we need to limit changes to machine deployments besides version changes if the control plane is upgrading (not scaling). From the machine deployment controller it looks like change to the version or any other change to the spec are treated the same way. So, is it okay to allow changes (not version) on a machine deployments when control plane is upgrading?  -- If we want to do this we might want to explicitly check if the control plane is upgrading (not necessarily scaling).

Reply:
I think you're right. Just thought about #5183 and it's very realistic that we want to handle scaling differently. Also a good point that we want to cover the edge cases explicitly.